File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head){
            return;
        }

        ListNode* fast = head;
        ListNode* slow = head;

        // 1 step
        while (fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }
        
        // 2 step
        ListNode* second = slow->next;
        slow->next = nullptr;
        ListNode* node = nullptr;

        while (second) {
            ListNode* temp = second->next;
            second->next = node;
            node = second;
            second = temp;
        }

        ListNode* first = head;
        second = node; // последний элемент списка

        while (second){
            ListNode* temp1 = first->next;
            ListNode* temp2 = second->next;

            first->next = second;
            second->next = temp1;

            first = temp1;
            second = temp2;
        }
    }
};

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def __init__(self):
        self.old_val_map = {}

    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return None

        if head in self.old_val_map:
            return self.old_val_map[head]
            
        copy = Node(head.val)
        self.old_val_map[head] = copy
        copy.next = self.copyRandomList(head.next)
        copy.random = self.old_val_map.get(head.random)
        return copy